packages/util/src/__specs__/polymorphism.spec.tsx

```diff
@@ -0,0 +1,107 @@
+/* eslint-disable @typescript-eslint/ban-ts-comment */
+
+import { render } from '@testing-library/react'
+import React, { MouseEvent } from 'react'
+
+import { forwardRefWithAs, forwardRefWithAsAndStatics } from '..'
+
+interface Props {
+  optional?: boolean
+}
+const Button = forwardRefWithAs<Props, 'button'>((props, ref) => {
+  const { as: Comp = 'button', optional, ...rest } = props
+  return <Comp data-optional={optional} ref={ref} {...rest} />
+})
+
+const Anchor = forwardRefWithAs<Props, 'a'>((props, ref) => {
+  const { as: Comp = 'a', optional, ...rest } = props
+  return <Comp data-optional={optional} ref={ref} {...rest} />
+})
+
+interface Statics {
+  Anchor: typeof Anchor
+  Button: typeof Button
+}
+
+const CompoundComp = forwardRefWithAsAndStatics<Props, 'div', Statics>(
+  (props, ref) => {
+    const { as: Comp = 'div', optional, ...rest } = props
+
+    return <Comp data-optional={optional} ref={ref} {...rest} />
+  }
+)
+CompoundComp.Anchor = Anchor
+CompoundComp.Button = Button
```

Comment:
What is being proven by doing this?

Reply:
that we can create polymorphic components that are also compound components. the new elements could replace the following code:

```
interface AvatarProps extends HTMLPropsFor<'div'> {
  ...
}

interface AvatarStatics {
  ...
}

type AvatarComponent = RefForwardingComponent<
  AvatarProps,
  HTMLDivElement,
  AvatarStatics
>

const Avatar = forwardRef<HTMLDivElement, AvatarProps>((props, ref) => {
  ...
}) as AvatarComponent
```

with this

```
interface AvatarProps extends HTMLPropsFor<'div'> {
  ...
}

interface AvatarStatics {
  ...
}

const Avatar = forwardRefWithAsAndStatics<AvatarProps, 'div', AvatarStatics>(
  (props, ref) => {
    ...
  }
)
```

The latter example requires less typing, no typecasting of the component, adds polymorphism

Reply:
i could also create a `forwardRefWithStatics` method if there is the need to create a compound component that doesn't support polymorphism. it could replace the `RefForwardingComponent` and would match this components in name and style